torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+        # For each node in path, which outputs were observed to be live
+        # before invoking graph recording, and after graph recording
+        self.recorded_liveness_before_graph: List[List[bool]] = []
+        self.recorded_liveness_after_graph: List[List[bool]] = []
+
+        # indices into node at level i for output j
+        self.expected_dead_indices_before_graph: List[Tuple[int, int]] = []
+        self.expected_dead_indices_after_graph: List[Tuple[int, int]] = []
+
+        # all live indices after graph recording
+        self.live_indices_after_graph: List[Tuple[int, int]] = []
+
+        if self.parent is not None:
+            previous_liveness = self.parent.recorded_liveness_after_graph
+            curr_liveness = self._get_liveness(self.path_weakrefs)
+
+            different_indices = self._get_different_indices(
+                previous_liveness, curr_liveness
+            )
+
+            self.recorded_liveness_before_graph = curr_liveness
+            self.expected_dead_indices_before_graph = different_indices
+
+        # graph needs to be kept alive for the duration of the model
+        # otherwise the memory pool would be freed on the first recording
+        inps_alloc_graph = torch.cuda.CUDAGraph()
+        recording_inputs = self._allocate_recording_inputs(inputs, inps_alloc_graph)
+
+        # graph used for recording model invocation
+        self.graph = torch.cuda.CUDAGraph()
+
+        # we allocate non-static inputs within the same memory pool as the CUDAGraph
+        # which we will record the model with. For memory efficiency, it is important
+        # to reclaim the input memory when the inputs are no longer live. To accomplish this,
+        # we record the metadata needed to reconstruct the inputs at their correct memory location,
+        # but do not keep them live during the cuda graph recording.
+        self.non_static_inputs_metadata = [
+            self._tensor_metadata(x) if idx not in (self.static_input_idxs) else None
+            for idx, x in enumerate(recording_inputs)
+        ]
+
+        stream = torch.cuda.Stream()
+        self._warmup(wrapped_function.model, stream, recording_inputs)
+
+        # on the first invocation, return the first recorded outputs, because their memory
+        # is correctly accounted for in the CUDAGraphs caching allocator, so on subsequent cudagraph
+        # recording we are tracing with a valid caching allocator state
+        self.recording_outputs = self._record(
+            wrapped_function.model, stream, recording_inputs
+        )
+
+        self.outputs_metadata = []
+
+        # As with inputs, we do not want to keep the outputs permanently alive because that would prevent
+        # their memory being reclaimed in subsequent cuda graph recordings. We record the tensor metadata
+        # needed to reconstruct instead.
+        for out in self.recording_outputs:
+            if isinstance(out, torch.Tensor):
+                self.device = (
```

Comment:
how would that work if function spans multiple devices? Should you assert that `out.device.index` is the same as `self.device`?

Reply:
I limit the cudagraphs with new api to a single device in a follow up pr, in the warm up one